File: src/mia/providers/deepseek.py
"""
DeepSeek Provider — 备选 LLM Provider

当 MiMo API 不可用时作为 fallback。
走标准 OpenAI 兼容协议。
"""

from typing import AsyncIterator, Optional

from loguru import logger
from openai import AsyncOpenAI

from mia.providers.base import BaseProvider


class DeepSeekProvider(BaseProvider):
    """DeepSeek API 封装 — OpenAI 兼容协议

    用于 Scheduler 和 TaskAgent 的备选 Provider。
    """

    BASE_URL = "https://api.deepseek.com/v1"
    CHAT_MODEL = "deepseek-chat"

    def __init__(
        self,
        api_key: str,
        base_url: Optional[str] = None,
    ):
        """
        Args:
            api_key: DeepSeek API Key (sk-xxxxx)
            base_url: 自定义 base URL
        """
        self.client = AsyncOpenAI(
            api_key=api_key,
            base_url=base_url or self.BASE_URL,
        )
        logger.info("DeepSeekProvider 初始化完成, base_url={}", base_url or self.BASE_URL)

    async def chat(
        self,
        messages: list[dict],
        model: Optional[str] = None,
        stream: bool = True,
        tools: Optional[list[dict]] = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
    ):
        """流式/非流式对话 — OpenAI 兼容"""
        kwargs = {
            "model": model or self.CHAT_MODEL,
            "messages": messages,
            "stream": stream,
            "max_tokens": max_tokens,
            "temperature": temperature,
        }
        if tools:
            kwargs["tools"] = tools
            kwargs["tool_choice"] = "auto"

        logger.debug(
            "DeepSeek chat: model={}, stream={}, msg_count={}",
            kwargs["model"], stream, len(messages),
        )
        return await self.client.chat.completions.create(**kwargs)

    async def chat_sync(
        self,
        messages: list[dict],
        model: Optional[str] = None,
        tools: Optional[list[dict]] = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
    ) -> str:
        """非流式对话 — 返回完整文本"""
        response = await self.chat(
            messages=messages,
            model=model,
            stream=False,
            tools=tools,
            max_tokens=max_tokens,
            temperature=temperature,
        )
        return response.choices[0].message.content or ""

    async def chat_stream(
        self,
        messages: list[dict],
        model: Optional[str] = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
    ) -> AsyncIterator[str]:
        """流式对话 — 逐 token 返回文本增量

        调用 self.chat(stream=True) 获取 OpenAI 兼容的 SSE 流，
        然后逐个 chunk 提取 delta.content 并 yield。
        """
        stream = await self.chat(
            messages=messages,
            model=model,
            stream=True,
            max_tokens=max_tokens,
            temperature=temperature,
        )
        async for chunk in stream:
            if chunk.choices and chunk.choices[0].delta.content:
                yield chunk.choices[0].delta.content

File: src/providers/deepseek.ts
/**
 * DeepSeek Provider — 备选 LLM Provider
 *
 * 当 MiMo API 不可用时作为 fallback。
 * 走标准 OpenAI 兼容协议。
 *
 * 与 Python 版 providers/deepseek.py 保持 1:1 语义映射。
 */

import OpenAI from 'openai';
import type { ChatCompletionMessageParam } from 'openai/resources/chat/completions';
import { BaseProvider } from './base.js';

/** DeepSeek API 默认配置 */
const BASE_URL = 'https://api.deepseek.com/v1';
const CHAT_MODEL = 'deepseek-chat';

/**
 * DeepSeekProvider — DeepSeek API 封装
 *
 * 用于 Scheduler 和 TaskAgent 的备选 Provider。
 */
export class DeepSeekProvider extends BaseProvider {
  private client: OpenAI;

  constructor(
    apiKey: string,
    baseUrl?: string,
  ) {
    super();
    this.client = new OpenAI({
      apiKey,
      baseURL: baseUrl || BASE_URL,
    });
  }

  /**
   * 流式/非流式对话 — OpenAI 兼容
   *
   * @returns stream=true → Stream; stream=false → ChatCompletion
   */
  async chat(
    messages: Array<{ role: string; content: unknown }>,
    model?: string,
    stream = true,
    tools?: Array<Record<string, unknown>>,
    maxTokens = 4096,
    temperature = 0.7,
  ): Promise<unknown> {
    const params: OpenAI.Chat.Completions.ChatCompletionCreateParams = {
      model: model || CHAT_MODEL,
      messages: messages as ChatCompletionMessageParam[],
      stream,
      max_tokens: maxTokens,
      temperature,
    };

    if (tools) {
      params.tools = tools as unknown as OpenAI.Chat.Completions.ChatCompletionTool[];
      params.tool_choice = 'auto';
    }

    return this.client.chat.completions.create(params, {
      ...(stream ? { stream: true as const } : { stream: false as const }),
    });
  }

  /** 非流式对话 — 返回完整文本 */
  async chatSync(
    messages: Array<{ role: string; content: unknown }>,
    model?: string,
    tools?: Array<Record<string, unknown>>,
    maxTokens = 4096,
    temperature = 0.7,
  ): Promise<string> {
    const response = (await this.chat(
      messages, model, false, tools, maxTokens, temperature,
    )) as OpenAI.Chat.Completions.ChatCompletion;

    return response.choices[0]?.message?.content || '';
  }

  /** 流式对话 — 逐 token 返回文本增量 */
  async *chatStream(
    messages: Array<{ role: string; content: unknown }>,
    model?: string,
    maxTokens = 4096,
    temperature = 0.7,
  ): AsyncGenerator<string, void, unknown> {
    const stream = (await this.chat(
      messages, model, true, undefined, maxTokens, temperature,
    )) as AsyncIterable<OpenAI.Chat.Completions.ChatCompletionChunk>;

    for await (const chunk of stream) {
      const delta = chunk.choices?.[0]?.delta?.content;
      if (delta) {
        yield delta;
      }
    }
  }
}
